jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java

```diff
@@ -509,13 +513,59 @@ protected void copyRequestHeaders(HttpServletRequest clientRequest, Request prox
 
     protected void addProxyHeaders(HttpServletRequest clientRequest, Request proxyRequest)
     {
-        addViaHeader(proxyRequest);
+        addViaHeader(clientRequest, proxyRequest);
         addXForwardedHeaders(clientRequest, proxyRequest);
     }
 
+    /**
+     * Adds the HTTP Via header to the proxied request.
+     *
+     * @deprecated Use {@link #addViaHeader(HttpServletRequest, Request)} instead.
+     * @param proxyRequest the request being proxied
+     */
+    @Deprecated
     protected void addViaHeader(Request proxyRequest)
     {
-        proxyRequest.header(HttpHeader.VIA, "http/1.1 " + getViaHost());
+        oldAddViaHeaderCalled = true;
+    }
+
+    /**
+     * Adds the HTTP Via header to the proxied request, taking into account data present in the client request.
+     * This method considers the protocol of the client request when forming the proxied request. If it
+     * is HTTP, then the protocol name will not be included in the Via header that is sent by the proxy, and only
+     * the protocol version will be sent. If it is not, the entire protocol (name and version) will be included. 
+     * If the client request includes a Via header, the result will be appended to that to form a chain.
+     *
+     * @param clientRequest the client request
+     * @param proxyRequest the request being proxied
+     * @see <a href="https://tools.ietf.org/html/rfc7230#section-5.7.1">RFC 7230 section 5.7.1</a>
+     */
+    protected void addViaHeader(HttpServletRequest clientRequest, Request proxyRequest)
+    {
+        // For backward compatibility reasons, call old, deprecated version of this method.
+        // If our flag isn't set, the deprecated method was overridden and we shouldn't do
+        // anything more.
+
+        oldAddViaHeaderCalled = false;
+        addViaHeader(proxyRequest);
+
+        if (!oldAddViaHeaderCalled)
```

Comment:
I don't like this test, pl;us I don't think it works if super is called.   I think you need to use reflection in `init()` to see if the method has been over ridden.     Then I think you should still do your checking for non http/1 proxy Request and only call the old method IF over ridden AND HTTP/1, otherwise always do it the new way.

Reply:
It works in this test at least:
```java
@Test
public void testInheritance() throws Exception
{
    ProxyServlet derivedProxyServlet = new ProxyServlet()
    {
        @Override
        protected void addViaHeader(Request proxyRequest)
        {
            System.err.println("addViaHeader called: " + proxyRequest);
            super.addViaHeader(proxyRequest);
        }
    };

    startServer(new HttpServlet()
    {
        @Override
        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            PrintWriter writer = resp.getWriter();
            writer.write(req.getHeader("Via"));
            writer.flush();
        }
    });
    String viaHost = "my-good-via-host.example.org";
    startProxy(derivedProxyServlet, Collections.singletonMap("viaHost", viaHost));
    startClient();

    HttpRequest proxyRequest = mockProxyRequest();
    derivedProxyServlet.addViaHeader(proxyRequest);

    ContentResponse response = client.GET("http://localhost:" + serverConnector.getLocalPort());

    assertThat("Response expected to contain content of Via Header from the request",
               response.getContentAsString(),
               Matchers.equalTo("1.1 " + viaHost));
}
```

This assumes we want to fix the other issues I mentioned that `http` as the protocol name was wrong and that we should not include protocol name nor the `/` when HTTP is used.

Reply:
Added this test an another that _does not_ call `super`. Can you recheck, @gregw ?

Reply:
Ah it is the other way around.  If somebody implements `addViaHeader(Request proxyRequest)` that doesn't call super, then you assume they have taken care of it.   OK.... I guess that works, but it is rather ugly.... hmm but so is reflection.

Reply:
> I guess that works, but it is rather ugly

It is certainly ugly. I think it wouldn't stick out as much if I would have done a check for the present of the `Via` header at this point, but that seemed to have a running time of _O(n)_. That's why I added this flag. It's a micro-optimization, but I'll take it.

Reply:
I don't think this can be fixed in general.
A subclass calling super and _expecting_ the header to be there will break (e.g. NPE).
Broken for broken, I'd say to implement the method empty, don't bother with the boolean flag and add the new logic.